worlds/marioland2/logic.py

```diff
@@ -0,0 +1,608 @@
+        reachable_coins += 3
+    if state.has("Space Physics", player):
+        reachable_coins += 79
+        if not auto_scroll:
+            reachable_coins += 21
+    return coins <= reachable_coins
+
+
+def space_zone_secret_course_coins(state, player, coins):
+    return coins <= 96 or not is_auto_scroll(state, player, "Space Zone Secret Course")
+
+
+def macro_zone_1_normal_exit(state, player):
+    return has_pipe_down(state, player) or state.has("Macro Zone 1 Midway Bell", player)
+
+
+def macro_zone_1_secret_exit(state, player):
+    return state.has("Fire Flower", player) and has_pipe_up(state, player) and macro_zone_1_midway_bell(state, player)
+
+
+def macro_zone_1_midway_bell(state, player):
+    return has_pipe_down(state, player) or state.has("Macro Zone 1 Midway Bell", player)
+
+
+def macro_zone_1_coins(state, player, coins):
+    auto_scroll = is_auto_scroll(state, player, "Macro Zone 1")
+    reachable_coins = 0
+    if has_pipe_down(state, player):
+        reachable_coins += 69
+        if auto_scroll:
+            if state.has_any(["Mushroom", "Fire Flower"], player):
+                reachable_coins += 5
+        else:
+            reachable_coins += 9
+            if state.has("Fire Flower", player):
+                reachable_coins += 19
+    elif state.has("Macro Zone 1 Midway Bell", player):
+        if auto_scroll:
+            reachable_coins += 16
+            if state.has_any(["Mushroom", "Fire Flower"], player):
+                reachable_coins += 5
+        else:
+            reachable_coins += 67
+    return coins <= reachable_coins
+
+
+def macro_zone_secret_course_coins(state, player, coins):
+    return state.has_any(["Mushroom", "Fire Flower"], player)
+
+
+def macro_zone_2_normal_exit(state, player):
+    return (has_pipe_down(state, player) or state.has("Macro Zone 2 Midway Bell", player)) and state.has(
+        "Water Physics", player) and has_pipe_up(state, player) and not is_auto_scroll(state, player, "Macro Zone 2")
+    
+    
+def macro_zone_2_midway_bell(state, player):
+    return ((has_pipe_down(state, player) and state.has("Water Physics", player))
+            or state.has("Macro Zone 2 Midway Bell", player))
+
+
+def macro_zone_2_coins(state, player, coins):
+    auto_scroll = is_auto_scroll(state, player, "Macro Zone 2")
+    if coins <= 27:
+        return True
+    if has_pipe_up(state, player) and state.has("Water Physics", player) and not auto_scroll:
+        if has_pipe_down(state, player):
+            return True
+        if state.has("Macro Zone 2 Midway Bell", player):
+            # Cannot return to the first section from the bell
+            return coins <= 42
+
+
+def macro_zone_3_normal_exit(state, player):
+    return ((has_pipe_down(state, player) and has_pipe_down(state, player))
+            or state.has("Macro Zone 3 Midway Bell", player),)
+
+
+def macro_zone_3_midway_bell(state, player):
```

Comment:
Since this logic seems identical to the `macro_zone_3_normal_exit` function, these two can be squashed together into one piece of logic

Reply:
The way these functions get called means this isn't an option

Reply:
Changed it